### DISCUSSION TITLE
Как прервать POST запрос и отдать клиенту сообщение об ошибке до того, как будут получены все данные?

### DISCUSSION THREAD

**Discussion by alexryndin** [Karlson2k/libmicrohttpd#17] [General]
Пример -- клиент объявил в заголовке Content-Length размер 1М, а отсылает вместо этого 100G. В идеале сервер должен иметь возможность отдать клиенту сообщение об ошибке, однако текущая версия, как я [понимаю](https://github.com/Karlson2k/libmicrohttpd/blob/a7c913850dd88d5ebb82e4b5013fdab1074dbaa8/src/microhttpd/connection.c#L5619), не позволяет добавлять ответ в очередь до тех пор, пока все байтики не будут получены сервером. Почему? Очевидно, что
1. В ситуации, когда клиент решил забить сервер байтиками, сервер не будет иметь возомжности красиво завершить соединение, ответив клиенту ошибкой (в противовес грубому разрыву)
2. Если во время итеративной обработки post запроса произошла ошибка, мы, опять же, будем вынуждены сначала пропустить все полученные данные, и только затем ответить клиенту.

**Comment by Karlson2k** [comment]
Текущая версия позволяет отправить ответ сразу после обработки заголовков или после получения полного тела запроса.
Естественно, если клиент сообщил, что размер тела составляет один мегабайт, то ровно этот мегабайт и будет прочитан в качестве тела запроса. Иначе не работает pipelining для запросов. Всё, что будет идти после мегабайта - будет обработано, как следующий запрос и, если это не запрос, то клиенту вернётся что-то с кодом 400, если на HTTP будет хоть как-то похоже.

В процессе чтения тела запроса в данный момент нельзя добавить ответ. Это связано с тем, что изначально это было сложно прикрутить ко всем режимам (а в режиме external polling ответ можно добавить и из другого потока). Но архитектура уже переписана заметно и в одной из ближайших версий добавим возможно ответа в любой момент времени. Если при этом тело запроса ещё не полносью прочиталось, то соединение будет принудительно закрытo после отправки ответа.

**Reply by alexryndin** [reply]
Благодарю. Насчёт неправильного Content-Length я ошибся -- действительно, libmicrohttpd в итоге будет обрабатывать дополнительные байты именно как другой запрос и вернёт 400...
А касательно ответа в любой момент -- это была бы очень полезная фича :) Спасибо за развитие проекта :)
